File: 2026/09-September/15-Tuesday/solution.py
class Solution:
    def maxPalindromes(self, s: str, k: int) -> int:
        n = len(s)

        # dp[i] = maximum number of valid non-overlapping
        # palindromes using the first i characters.
        dp = [0] * (n + 1)

        # Expand around every possible center.
        for center in range(n):
            # Odd-length palindrome
            l = r = center
            while l >= 0 and r < n and s[l] == s[r]:
                if r - l + 1 >= k:
                    dp[r + 1] = max(dp[r + 1], dp[l] + 1)
                l -= 1
                r += 1

            # Even-length palindrome
            l, r = center, center + 1
            while l >= 0 and r < n and s[l] == s[r]:
                if r - l + 1 >= k:
                    dp[r + 1] = max(dp[r + 1], dp[l] + 1)
                l -= 1
                r += 1

        # Carry forward the best answer when we don't
        # end a palindrome at position i.
        for i in range(1, n + 1):
            dp[i] = max(dp[i], dp[i - 1])

        return dp[n]

File: 2026/09-September/15-Tuesday/solution.cpp
class Solution {
public:
    int maxPalindromes(string s, int k) {
        int n = s.size();

        // isPal[i][j] = true if s[i...j] is a palindrome
        vector<vector<bool>> isPal(n, vector<bool>(n, false));

        // Build palindrome table
        for (int i = n - 1; i >= 0; i--) {
            for (int j = i; j < n; j++) {
                if (s[i] == s[j] &&
                    (j - i <= 2 || isPal[i + 1][j - 1])) {
                    isPal[i][j] = true;
                }
            }
        }

        // dp[i] = maximum number of non-overlapping
        // palindromes using the first i characters
        vector<int> dp(n + 1, 0);

        for (int end = 1; end <= n; end++) {
            // Skip current character
            dp[end] = dp[end - 1];

            // Try every palindrome ending at end - 1
            for (int start = 0; start <= end - k; start++) {
                if (isPal[start][end - 1]) {
                    dp[end] = max(dp[end], dp[start] + 1);
                }
            }
        }

        return dp[n];
    }
};
